### PR TITLE
Hetzner remote provison

### PR DESCRIPTION
**Pull request by ableischwitz**
## Description
Re-worked Hetzner deployment, renamed Ansible modules

## Checklist/ToDo's

- [X] Added documentation?
- [X] Added release note entry? (  docs/release-notes.md )
- [X] Tested?

### CODE REVIEW

**Review by rbo — CHANGES_REQUESTED** [review]
I merged the PR ##225  with a bunch of ansible module changes. I guess there some duplications here. 

It's possible to split the changes a bit? 

What was the thought behind the change? Running the playbook complete from your workstation?

**Comment by rbo** [comment]
Maybe it makes sense to rebase from devel branch again.

File: ansible/setup.yml [hunk 1]
```diff
@@ -1,3 +1,4 @@
 ---
-- import_playbook: 01-prepare-host.yml
-- import_playbook: 02-create-cluster.yml
+- ansible.builtin.import_playbook: 00-provision-hetzner.yml
```
[the remark this answers is not in the document]
Comment (ableischwitz) [reply ansible/setup.yml:2]: The reason for running the provisioning too was that I didn't want (wasn't able) to clone the github repos on a remote system.
As my Hetzner system is IPv6 only, there is no easy way to connect to a IPv4 only service like github.com. Also the philosophy behind Ansible suggests to have a bastion host which should not be the host to be managed. Using a external system (like my work-horse) just makes a lot more sense.

**Comment by rbo** [comment]
I cherry-picked some commits from you @ableischwitz and applied into branch [pr226](https://github.com/RedHat-EMEA-SSA-Team/hetzner-ocp4/tree/pr226) . 
It worked very well,  you have to change the `ansible_host` in `inventory/hosts.yaml`.

The only issue I/we have to fix is the "artifacts" directory it creates at remote site, the same directory structure where your repo is clone locally.:
```
$ ssh -l root tester.openshift.pub tree /var/home/rbohne/Devel/github.com/RedHat-EMEA-SSA-Team/hetzner-ocp4/
Warning: Permanently added 'tester.openshift.pub' (ED25519) to the list of known hosts.
/var/home/rbohne/Devel/github.com/RedHat-EMEA-SSA-Team/hetzner-ocp4/
├── ansible
├── certificate
│   ├── account.key
│   ├── isrgrootx1.pem
│   └── tester.openshift.pub
│       ├── ca-bundle.pem
│       ├── cert.crt
│       ├── cert.csr
│       ├── cert.key
│       ├── fullchain.crt
│       └── intermediate.crt
└── tester
    ├── auth
    │   ├── kubeadmin-password
    │   └── kubeconfig
    ├── bootstrap-extra.ign
    ├── bootstrap.ign
    ├── bootstrap.rcc
    ├── install-config.yaml.original
    ├── master-extra.ign
    ├── master.ign
    ├── master.rcc
    ├── metadata.json
    ├── worker-extra.ign
    ├── worker.ign
    └── worker.rcc

5 directories, 21 files
``

**Comment by rbo** [comment]
It's not that easy, just to recap the overview:
```
  ┌───────────────────────────────────┐
  │                                   │
  │ Hetzner Server                 ▲  │
  │                                │  │
  └────────────────────────────────┼──┘
                                   │
                                  SSH
                                   │
  ┌────────────────────────────────┬───┐
  │Local Workstation/Laptop        │   │
  │                                │   │
  │ ┌──────────────────────────────┼─┐ │
  │ │ Ansible runner (Podman)      │ │ │
  │ │                                │ │
  │ │                                │ │
  │ └────────────────────────────────┘ │
  │                                    │
  └────────────────────────────────────┘

```

Most of the artifacts (certificates, openshift ignition, auth/ folder,..) has to be stored and used at the Hetzner Server

Some of the modules/tasks have to be run in the Ansible runner because in the runner all modules are installed. 
That's why you installed the packages at the machine: https://github.com/ableischwitz/hetzner-ocp4/blob/9055e45e702673dad54f8939714ff0731c7fc409/ansible/roles/provision-hetzner/tasks/main.yml#L49-L64 ?

At the end, we have to decide:

**first)** which task has to be run where (ansible runner or hetzner server) and
  * most of the task are executed at the hetzner server except:
    * API Calls to dns provider
    * module `kubernetes.core.k8s`   

**second)** where the data has to be stored (the location at the hetzner server) OR even on local workstation. 
  * I suggest storing all generate data at the hetzner server (letsencrypt certificates & openshift installer artifacts) 
  * SSH Key for coreos, default `~/.ssh/...` means local laptop or hetzner server, depending on where the ansible runner is running. 
      suggestion: always from the target machine. 
  * From local machine (where the ansible runner is running) we fetch only the `cluster.yml`.

**Comment by rbo** [comment]
So, I made some changes in my branch [pr226](https://github.com/RedHat-EMEA-SSA-Team/hetzner-ocp4/tree/pr226) - it looks good....

**Comment by rbo** [comment]
One issue I have:
```
[root@tester ~]# showmount -e 192.168.52.1
rpc mount export: RPC: Unable to receive; errno = Connection refused
[root@tester ~]# 
```

After a reboot NFS works fine... 

```
[root@tester ~]# systemctl status rpcbind
● rpcbind.service - RPC Bind
   Loaded: loaded (/usr/lib/systemd/system/rpcbind.service; enabled; vendor preset: enabled)
   Active: active (running) since Sun 2022-11-27 11:30:49 CET; 3h 51min ago
     Docs: man:rpcbind(8)
 Main PID: 56302 (rpcbind)
    Tasks: 1 (limit: 822480)
   Memory: 1.8M
   CGroup: /system.slice/rpcbind.service
           └─56302 /usr/bin/rpcbind -w -f

Nov 27 11:30:49 tester systemd[1]: Starting RPC Bind...
Nov 27 11:30:49 tester systemd[1]: Started RPC Bind.
[root@tester ~]# systemctl status nfs-server
● nfs-server.service - NFS server and services
   Loaded: loaded (/usr/lib/systemd/system/nfs-server.service; enabled; vendor preset: disabled)
   Active: active (exited) since Sun 2022-11-27 11:30:51 CET; 3h 51min ago
  Process: 56515 ExecStart=/bin/sh -c if systemctl -q is-active gssproxy; then systemctl reload gssproxy ; fi (code=exited, >
  Process: 56500 ExecStart=/usr/sbin/rpc.nfsd (code=exited, status=0/SUCCESS)
  Process: 56497 ExecStartPre=/usr/sbin/exportfs -r (code=exited, status=0/SUCCESS)
 Main PID: 56515 (code=exited, status=0/SUCCESS)
    Tasks: 0 (limit: 822480)
   Memory: 0B
   CGroup: /system.slice/nfs-server.service

Nov 27 11:30:51 tester systemd[1]: Starting NFS server and services...
Nov 27 11:30:51 tester systemd[1]: Started NFS server and services.
[root@tester ~]# 

```

**Comment by rbo** [comment]
https://access.redhat.com/solutions/6973007 

We have to boot into the new kernel
```
[root@tester ~]# rpm -qa | grep kernel
kernel-modules-4.18.0-147.el8.x86_64
kernel-core-4.18.0-147.el8.x86_64
kernel-4.18.0-147.el8.x86_64
kernel-tools-4.18.0-425.3.1.el8.x86_64
kernel-tools-libs-4.18.0-425.3.1.el8.x86_64
kernel-core-4.18.0-425.3.1.el8.x86_64
kernel-modules-4.18.0-425.3.1.el8.x86_64
kernel-4.18.0-425.3.1.el8.x86_64
[root@tester ~]# 
```

**Comment by rbo** [comment]
One shot reset the server install rhel8, add subscription install ocp4 🤩 
```
PLAY RECAP *********************************************************************
host                       : ok=191  changed=118  unreachable=0    failed=0    skipped=66   rescued=0    ignored=1   
Monday 28 November 2022  18:38:44 +0000 (0:00:00.035)       0:43:57.987 ******* 
=============================================================================== 
openshift-4-cluster : Waiting for bootstrap to complete --------------- 611.04s
openshift-4-cluster : Waiting installation to complete ---------------- 503.68s
openshift-4-cluster : Upgrade all packages ---------------------------- 353.25s
provision-hetzner : Wait 600 seconds for port 22 to become open ------- 199.45s
openshift-4-cluster : Wait for the reboot to complete ----------------- 173.83s
openshift-4-cluster : Installing KVM Packages ------------------------- 131.31s
provision-hetzner : Run installimage ---------------------------------- 113.07s
openshift-4-cluster : RHEL Subscription -------------------------------- 81.64s
openshift-4-cluster : Disable all RHSM repositories -------------------- 63.62s
openshift-4-cluster : Enable repos for RHEL 8 -------------------------- 55.01s
provision-hetzner : Wait 600 seconds for port 22 to become open -------- 28.42s
openshift-4-loadbalancer : Install haproxy ----------------------------- 20.39s
openshift-4-cluster : Download Openshift installer --------------------- 18.12s
openshift-4-cluster : download coreos ---------------------------------- 16.91s
openshift-4-cluster : unzip the coreos --------------------------------- 12.59s
public_dns : Create Cloudflare DNS records ----------------------------- 11.07s
openshift-4-cluster : Copy/dd coreos into device ----------------------- 11.01s
openshift-4-cluster : Copy/dd coreos into device ----------------------- 10.90s
openshift-4-cluster : Copy/dd coreos into device ----------------------- 10.83s
openshift-4-cluster : Copy/dd coreos into device ----------------------- 10.75s
```

**Comment by rbo** [comment]
partially merged with #226

**Pull request closed without merging**